src/motor/motor_ip.cpp

```diff
@@ -0,0 +1,149 @@
+#include "motor_ip.h"
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+#include <arpa/inet.h>
+#include <errno.h>
+#include "poll.h"
+
+namespace obot {
+
+void UDPFile::open() {
+    fd_ = ::socket(AF_INET, SOCK_DGRAM, 0);
+    addrinfo hints = {};
+    hints.ai_family = AF_INET;      // IPv4
+    hints.ai_socktype = SOCK_DGRAM; // UDP
+
+    addrinfo *result, *res;
+    int addr_info_result = getaddrinfo(ip_.c_str(), port_.c_str(), &hints, &result);
+    if (addr_info_result != 0) {
+      throw std::runtime_error("getaddrinfo failed for " + ip_ + ":" + port_ + ", error: " + std::to_string(addr_info_result));
+    }
+
+    char addrstr[100];
+    void *ptr;
+    res = result;
+    int n_results = 0;
+    while(res) {
+        n_results++;
+        socklen_t addr_len = 0;
+        if (res->ai_family != AF_INET) {
+          throw std::runtime_error(ip_ + ":" + port_ + ", not ipv4");
+        }
+        ptr = &((struct sockaddr_in *) res->ai_addr)->sin_addr;
+        addr_len = sizeof(sockaddr_in);
+        inet_ntop (res->ai_family, ptr, addrstr, 100);
+        addrstr_ = addrstr;
+        res = res->ai_next;
+    }
+    if (n_results != 1) {
+      throw std::runtime_error(ip_ + ":" + port_ + ", n_results error: " + std::to_string(n_results));
+    }
+    std::memcpy(&addr_, result->ai_addr, result->ai_addrlen);
+    freeaddrinfo(result);
+    flush();
+}
+
+void UDPFile::flush() {
+    int result;
+    char c[1024];
+    do {
+      result = read(c, 1024);
+    } while (result > 0);
+}
+
+int UDPFile::poll() {
+    pollfd tmp;
+    tmp.fd = fd_;
+    tmp.events = POLLIN;
+    int poll_result = ::poll(&tmp, 1, timeout_ms_);
+    if (poll_result == 0) {
+        return -ETIMEDOUT;
+    } else if (poll_result < 0) {
+        return poll_result;
+    }
+    return poll_result;
+}
+
+ssize_t UDPFile::read(char * data, unsigned int length) {
+    ObotPacket packet;
+    packet.command = 2;
+    packet.mailbox = recv_mailbox_;
+    int send_result = sendto(fd_, &packet, 6, 0, (sockaddr *) &addr_, sizeof(addr_));
```

Comment:
Can we add a ObotPacket::Header with the OBOT, command and mailbox then use sizeof(ObotPacket::Header) here and below